### PR TITLE
Dev - Remove support of secrets.json and other improvements #166

### PR DESCRIPTION
**Pull request by stanfrbd**
## Summary
- Completes the transition away from `secrets.json` toward `.env` / environment-variable-only configuration.
- Removes the unauthenticated `/config` GUI endpoint.
- Adds a `secrets.json` → `.env` migration script with safe value quoting and optional sample validation.
- Updates docs, Docker image tagging, and CI triggers.

**What's included:**
- Uses only `.env` for configuration
- No more `/config` endpoint
- Adds PR template
- Triggers build workflows only on tag `v*`
- Adds release tags for images (`latest` / `v0.x.x`)
- Adds migration script from `secrets.json` to `.env`
- `--secrets-sample` argument in migration script is now optional (skips `validate_samples` when not provided, since `secrets-sample.json` was removed)
- Migration script now quotes/escapes `.env` values containing `#`, surrounding whitespace, backslashes, embedded quotes, or newlines to ensure safe round-tripping with python-dotenv
- Fixed typo in `utils/config.py` docstring (`envrionment` → `environment`)
- Clarified `load_env_file` log messages: `.env` presence logged at `INFO` as optional/local-dev; missing `.env` demoted to `DEBUG` (no false warning in production deployments using injected env vars)
- Fixed `read_secrets_from_env` docstring to remove stale reference to "config file"
- Fixed MkDocs admonition indentation in `Advanced-options-for-deployment.md`
- Replaced tab indentation with spaces in `Reload-secrets-and-configuration.md` list-embedded code block
- Hardened `prod/bootstrap.sh`: replaced conditional `rm` with `rm -f` to avoid aborting container startup under `set -e` when the version cache file is missing or unremovable
- Updated `Reload-secrets-and-configuration.md` and `Upgrade-Cyberbro.md` "Without docker" sections to remove stale `supervisord` references; now recommend restarting the `gunicorn` process/service via systemd or a generic process manager
- Fixed `Advanced-options-for-deployment.md` runtime loading behavior section to accurately describe the `DEBUG`-level (not warning) log message emitted when `.env` is absent
- Fixed `MDEEngine` class-level token cache inconsistency: `_get_token()` now uses `type(self)._cached_token` for writes, making reads and writes consistently target the class-level attribute and preventing shadowing instance attributes

## Scope
- [ ] Bug fix
- [x] New feature
- [x] Refactor
- [x] Documentation
- [ ] New engine
- [ ] Other (describe):

## Validation (required)
- [x] I ran relevant tests locally and they pass.
- [x] I validated real behavior (not only code style or static checks).
- [x] I checked edge cases and failure paths.

### Test evidence

```bash
# Linter
python -m ruff check scripts/secrets_json_to_env.py utils/config.py engines/microsoft_defender_for_endpoint.py
# All checks passed

# Migration script runs without --secrets-sample (no FileNotFoundError)
python scripts/secrets_json_to_env.py --secrets secrets.json --output .env

# Quoting behaviour verified manually:
# - values with '#', leading/trailing spaces, backslashes, quotes, newlines → double-quoted with escaping
# - simple API key values → left unquoted for readability

# MDE engine tests (67 tests)
pytest tests/engines/test_microsoft_defender_for_endpoint.py -v
# 67 passed
```

## AI-assisted contribution disclosure
- [x] This PR includes AI-assisted work.
- [x] I reviewed and understood all generated code.
- [x] I refined AI output where needed (no low-quality slop).
- [x] I am fully responsible for this submission.

If AI was used, briefly describe what parts were AI-assisted:
Migration script refactoring, optional argument handling, value quoting/escaping logic, docstring fixes, log message rewording, docs indentation fixes, bootstrap hardening, supervisord reference removal, docs log-level accuracy fix, and MDEEngine cache consistency fix were AI-assisted.

## Maintainer merge policy acknowledgement
- [x] I understand uncertain changes may be merged to `dev` first for additional validation.
- [x] I understand `main` is kept clean/stable for releases and tags.

### Engine Implementation
- [ ] Added engine file in `engines/engine_name.py` (using existing engines as template).
- [ ] Added engine import/registration in `engines/__init__.py`.
- [ ] Added engine config/secret variable in `utils/config.py` (if relevant).

### Configuration & Secrets
- [ ] Added needed API key/config vars to `.env.sample`.
- [ ] Verified variable templating in `docker-compose.yml`.

### UI & Frontend
- [ ] Engine result is copyable via GUI in `static/format_results.js`.
- [ ] Added/updated layouts in `templates/engines_layouts/` for card/table rendering.
- [ ] Added engine in `templates/display_cards.html` and `templates/display_table.html`.
- [ ] Added engine in `templates/index.html` with description (alphabetic order).
- [ ] Added engine support for graph view in `templates/graph.html`.

### Documentation
- [ ] Added docs in `docs/api-keys/Get-Engine-API-key.md` (if relevant).
- [ ] Added API key guide link in `docs/index.md` (if relevant).
- [ ] Added engine to `docs/quick-start/API-usage-and-engine-names.md`.
- [ ] Added page in `mkdocs.yml` sidebar.
- [ ] Updated `docs/quick-start/Quick-start-&-Installation.md` and `.env` examples.
- [ ] Added env vars to `docs/quick-start/Advanced-options-for-deployment.md` (`docker compose` example).
- [ ] Updated `README.md` references (env vars and API/services section).

## Risk and rollback
- Risk level: Low
- Potential impact: Users still relying on `secrets.json` must migrate to `.env` using the provided migration script. Migrated `.env` files now correctly quote values containing special characters, preventing silent data loss during migration. The `MDEEngine` token cache fix is a correctness improvement with no functional regression.
- Rollback strategy: Revert to previous commit and restore `secrets.json` support.

## Additional notes
- TODO: Add SOPS tutorial / docs
- Nice to have (future PR): Add native support for Azure Key Vault

### CODE REVIEW

File: .github/workflows/jobs.yml [hunk 1]
```diff
@@ -2,6 +2,7 @@ name: Build and Test
 
 on:
   push:
+  pull_request:
   workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply .github/workflows/jobs.yml:6]: Only the build workflows are affected, this one is the tests check workflow

File: scripts/secrets_json_to_env.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+def parse_args() -> argparse.Namespace:
+    """Parse command line arguments."""
+
+    parser = argparse.ArgumentParser(description="Convert secrets.json into a .env file.")
+    parser.add_argument("--secrets", type=Path, default=Path("secrets.json"))
+    parser.add_argument("--secrets-sample", type=Path, default=Path("secrets-sample.json"))
+    parser.add_argument("--env-sample", type=Path, default=Path(".env.sample"))
+    parser.add_argument("--output", type=Path, default=Path(".env"))
+    return parser.parse_args()
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply scripts/secrets_json_to_env.py:38]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3027886098) directly in this branch

File: utils/config.py [hunk 3]
```diff
@@ -182,7 +163,7 @@ def read_secrets_from_env(secrets: Secrets) -> Secrets:
     Override the config file if the environment variable is set.
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply utils/config.py]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3027886138)

File: scripts/secrets_json_to_env.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        "",
+    ]
+    used_keys: set[str] = set()
+
+    for key in ordered_keys:
+        value: str | None = env_values.get(key)
+        if value is None or value == "":
+            continue
+        rendered.append(f"{key}={value}")
+        used_keys.add(key)
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply scripts/secrets_json_to_env.py:132]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3036823234)

File: docs/quick-start/Advanced-options-for-deployment.md [hunk 5]
```diff
@@ -134,273 +169,88 @@ services:
...
-    In Docker, you can also set `HOST_PORT` (default `5000`) for the host-side mapping.
-
-## API prefix in `app.py` and `index.html` options
-
-**In `secrets.json`:**
-
-!!! tip
-    By default, the API is accessible at `http://cyberbro_instance:5000/api`
+    In Docker, keep `FLASK_PORT` and `HOST_PORT` aligned with your `ports:` mapping.
+  `FLASK_HOST` is forced to `0.0.0.0` in the container so a local `127.0.0.1` from your shell or `.env` cannot break exposure.
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply docs/quick-start/Advanced-options-for-deployment.md]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3036823246)

File: docs/quick-start/Reload-secrets-and-configuration.md [hunk 6]
```diff
@@ -1,18 +1,12 @@
...
-5. Then go to http://127.0.0.1:5000 and use Cyberbro.
-
-### The container cyberbro is not running
-1. Edit `secrets.json` / `.env` with the updated values.
-2. ```docker compose up --build --force-recreate -d```
+1. Edit `.env` on the host machine (next to `docker-compose.yml`).
+2. Recreate the container so new environment variables are injected:
+	```bash
+	docker compose up --build --force-recreate -d
+	```
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply docs/quick-start/Reload-secrets-and-configuration.md]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3036823251)

File: prod/bootstrap.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+if [ "$#" -eq 0 ]; then
+    echo "[bootstrap] error: no startup command provided"
+    exit 1
+fi
+
+# Remove stale version cache after upgrades so update checks are refreshed.
+if [ -f "data/version_cache.json" ]; then
+    rm "data/version_cache.json"
+fi
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply prod/bootstrap.sh]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3036823274)

File: utils/config.py [hunk 8]
```diff
@@ -140,49 +140,31 @@ def update(self, updated_secrets: dict[str, Any]) -> None:
...
+    if env_file.exists():
+        load_dotenv(dotenv_path=env_file, override=False)
+        logger.info("Loaded environment variables from .env file. Don't use in production.")
+        return
 
-    return secrets
+    logger.warning(
+        "'.env' not found at %s (expected in production). Falling back to environment variables.",
+        env_file,
+    )
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply utils/config.py:163]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3036823240)

File: docs/quick-start/Reload-secrets-and-configuration.md [hunk 9]
```diff
@@ -1,18 +1,12 @@
...
+    ```bash
+    docker compose up --build --force-recreate -d
+    ```
 3. Then go to http://127.0.0.1:5000 and use Cyberbro.
 
 ## Without docker
-1. Edit `secrets.json` with the updated values.
+1. Edit `.env` (or exported environment variables) with the updated values.
 2. Restart your `gunicorn` instance or start `supervisord` e.g. `service supervisord restart`
 3. Then go to http://127.0.0.1:5000 and use Cyberbro.
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply docs/quick-start/Reload-secrets-and-configuration.md:12]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3036878934)

File: engines/microsoft_defender_for_endpoint.py [hunk 10]
```diff
@@ -82,8 +87,10 @@ def _get_token(self) -> str:
 
         try:
             aad_token = json_response["access_token"]
-            token_path = Path("mde_token.txt")
-            token_path.write_text(aad_token)
+
+            with self._token_lock:
+                self._cached_token = aad_token
+
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply engines/microsoft_defender_for_endpoint.py:93]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3036878951)

File: docs/quick-start/Advanced-options-for-deployment.md [hunk 11]
```diff
@@ -1,19 +1,59 @@
...
+
+```bash
+docker compose up --build
+```
+
+## Runtime loading behavior
+
+At startup, Cyberbro tries to load `.env` from the project root.
+If `.env` is missing, Cyberbro falls back to process environment variables and logs a warning.
+
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply docs/quick-start/Advanced-options-for-deployment.md:30]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3036878925)

File: docs/quick-start/Upgrade-Cyberbro.md [hunk 12]
```diff
@@ -37,4 +37,4 @@ docker-compose up -d --pull always --force-recreate
 * `gunicorn -b 0.0.0.0:5000 app:app --timeout 120` (or using `supervisord.conf`)
```
[the remark this answers is not in the document]
Comment (stanfrbd) [reply docs/quick-start/Upgrade-Cyberbro.md:38]: @copilot apply changes based on [this feedback](https://github.com/stanfrbd/cyberbro/pull/170#discussion_r3036878941)
